File: src/algorithm/lesson_4_reduction/ReductionDynamic/lomuto_And_quickSelect/quickSelect.cc
#include "lomutoPartition.h"
#include "quickSelect.h"
/**
 * 快速选择是一种从无序列表找到第k小元素的选择算法。
 * 具有很好的平均时间复杂度，然而最坏时间复杂度则不理想.
 * 性能： 从O(n log n)至O(n)，不过最坏情况仍然是O(n2)。
*/
int quickSelect(int a[], int len, int k)
{
    int l = 0, r = len - 1;

    int partitionIdx = lomutoPartition(a, l, r);

    while (l <= r)
    {
        if ((k - 1) == partitionIdx)
        {
            break;
        }
        else if ((k - 1) < partitionIdx)
        {
            r = partitionIdx - 1;
        }
        else
        {
            l = partitionIdx + 1;
        }

        partitionIdx = lomutoPartition(a, l, r);
    }
    return a[partitionIdx];
}

File: src/algorithm/algorithm_design/2.Reduce-and-Conquer/减可变规模/select/quickSelect.py
from LomutoPartition import lomutoPartition


#
#   快速选择 是一种从无序列表找到第k小元素的选择算法。
#   具有很好的平均时间复杂度，然而最坏时间复杂度则不理想.
#   性能： 从O(n log n)至O(n)，不过最坏情况仍然是O(n2)。
#
def quickSelect(a: list, k: int):
    s = lomutoPartition(a)
    if s == (k - 1):
        return a[s]
    elif s > (k - 1):
        return quickSelect(a[:s], k)
    else:
        return quickSelect(a[s+1:], k-1-s)